# Assignment 2: Obesity Diagnosis based on eating Habits and physical condition

Student Name: Xing Yang Goh

Student ID: 1001969

#### <b>Due date</b>: Monday, 6 September 2021 5pm

<b>Submission method</b>: Canvas submission

<b>Submission materials</b>: completed copy of this iPython notebook

<b>Late submissions</b>: -10% per day (both week and weekend days counted up to 5 days)
<ul>
    <li>one day late, -1.5;</li>
    <li>two days late, -3.0;</li>
    <li>three days late, -4.5;</li>
    <li>four days late, -6.0;</li>
    <li>five days late, -7.5;</li>
</ul>

<b>Marks</b>: 15% of mark for class. 


<b>Materials</b>: See [Using Jupyter Notebook and Python page](https://canvas.lms.unimelb.edu.au/courses/105477/pages/python-and-jupyter-notebooks?module_item_id=2613813) on Canvas (under Modules> Coding Resources) for information on the basic setup required for this class, including an iPython notebook viewer and the python packages Numpy, Scipy, Scikit-Learn. The instructions for each question will state what libraries you can use; if your iPython notebook doesn't run on the marker's machine, you will lose marks. <b> You should use Python 3</b>.  


<b>Evaluation</b>: Your iPython notebook should run end-to-end without any errors in a reasonable amount of time, and you must follow all instructions provided below, including specific implementation requirements and instructions for what needs to be printed (please avoid printing output we don't ask for). You should edit the sections below where requested, but leave the rest of the code as is. You should leave the output from running your code in the iPython notebook you submit, to assist with marking. The amount each section is worth is given in parenthesis after the instructions. 

You should be careful to use Python built-in functions and operators when appropriate and pick descriptive variable names that adhere to <a href="https://www.python.org/dev/peps/pep-0008/">Python style requirements</a>. If you think it might be unclear what you are doing, you should comment your code to help the marker make sense of it. While the main focus is on correctness of your methods, you will lose marks if your code is not understandable.

<b>Updates</b>: Any major changes to the assignment will be announced via Canvas. Minor changes and clarifications will be announced on the discussion board; we recommend you check it regularly.

<b>Academic misconduct</b>: For most people, collaboration will form a natural part of the undertaking of this homework, and we encourage you to discuss it in general terms with other students. However, this ultimately is still an individual task, and so reuse of code or other instances of clear influence will be considered cheating. We will be checking submissions for originality and will invoke the University’s <a href="http://academichonesty.unimelb.edu.au/policy.html">Academic Misconduct policy</a> where inappropriate levels of collusion or plagiarism are deemed to have taken place.

## Overview

In this assignment, you will be working with *obesity* dataset containing data for the estimation of obesity levels in individuals from the countries of Mexico, Peru and Colombia, based on their eating habits and physical condition. The data contains 16 attributes describing 2111 patients. The dataset provided for this assignment is derived from <a href="https://archive.ics.uci.edu/ml/datasets/Estimation+of+obesity+levels+based+on+eating+habits+and+physical+condition+#">this</a> resource. Your task is to predict whether a patient is considered obese (mass body index >=30) or not (mass body index <30).

The attributes are

|ID|Feature Name| Feature Type | Feature Values|
| :-| :-| :-| :-|
|0|Gender| categorical | Female: 0, Male: 1|
|1|Age| continuous | |
|2|Height| continuous | |
|3|Weight| continuous | ? |
|4|family_history| categorical | yes: 1, no: 0|
|5|FACV| categorical | yes: 1, no: 0|
|6|FCVC| continuous | |
|7|NCP| continuous | |
|8|CAEC| categorical | Always: 0, Frequently: 1, Sometimes: 2, no:  3|
|9|SMOKE| categorical | yes: 1, no: 0|
|10|CH2O| continuous | |
|11|SCC| categorical | yes: 1, no: 0|
|12|FAF| continuous | |
|13|TUE| continuous | |
|14|CALC| categorical | Always: 0, Frequently: 1, Sometimes: 2, no:  3, ?|
|15|MTRANS| categorical | Automobile: 0, Bike: 1, Motorbike: 2, Public_Transportation: 3, Walking: 4, ?|

You can find out more about the individual attributes / values, and the origin of the data set in this <a href="https://www.sciencedirect.com/science/article/pii/S2352340919306985"> article</a>  


You will build a number of classifiers to predict whether a patient is obese or not based on the attributes above.


#### The following instructions hold for every question in the assignment
- leave the order of instances intact, i.e., do not shuffle the data
- do not change the names or types of variables provided by us in the code cells below
- '?' denotes a missing value, and you should treat it as instructed in Question 1-B.
- the mapping of categorical feature values to integers indicates how you should process your data in 1-C.

## Question 1: Loading and pre-processing the data [2 marks]

You were provided with two data files:

**obesity.train** contains about 1,688 training instances, one instance per line in comma-separated value (csv) format. Each line contains 17 fields. The first 16 fields correspond to the features listed above, the final field denotes the class label

**obesity.test** is formatted exactly like obesity.train, and contains about 423 further instances for evaluation.

### 1-A Read the data [0.25 marks]

First, you will read in the data and create train features, train labels, test features and test labels. Do not apply any data transformations in this step and keep the '?' as is. 

In [1]:
# load data

data = open("obesity.train",'r').readlines()
test_data = open("obesity.test",'r').readlines()
feature_names = ["gender","age","height","weight","family_history","FACV","FCVC","NCP","CAEC","smoke","CH2O","SCC","FAF","TUE","CALC","MTRANS"]
feature_index = [0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15]

x_train = []
y_train = []
x_test = []
y_test = []

# split the training set into feature instances and class labels
for line in data:
    line = line.strip() #remove all leading and trailing whitespace (i.e., the newline symbol '\n')
    line = line.split(',')
    featureInst = line[0:16] #ignore the unique identifier
    classLabel = line[16]
    y_train.append(classLabel)
    x_train.append(featureInst)

# split the testing set into feature instances and class labels
for line in test_data:
    line = line.strip() #remove all leading and trailing whitespace (i.e., the newline symbol '\n')
    line = line.split(',')
    featureInst = line[0:16] #ignore the unique identifier
    classLabel = line[16]
    y_test.append(classLabel)
    x_test.append(featureInst)

<b>For your testing:</b>

In [2]:
assert len(x_train)==len(y_train)==1688
assert len(x_train[0])==len(x_test[0])==16
assert len(x_test)==len(y_test)==423
assert x_train[5][2] == '1.7544389999999999'
assert x_test[8][6] == '2.0'

### 1-B: Dealing with Missing Values [0.75 marks]

You will write <b>two functions</b> to replace missing values using the strategies below based on the different attribute types (categorical, continuous):

   * <b>Continuous features</b>: For each feature find the <b>average feature value</b> in the train dataset 
   * <b>Categorical features</b>: For each feature find the <b>most frequent value</b> in the train dataset 

and a <b>third function</b> that replaces the missing values in the train and test datasets with the most frequent (categorical features) and average values (continuous features) that you found in the train dataset.
 

**Your tasks**

You will implement three functions
1. `num_mean_feat`
<ul>
  <li>input:  instances with only continuous features</li>
  <li>output: a list containing the average value for each continuous feature (aka substitute values)</li>
</ul> 
    
2. `cat_most_feat`
<ul>
    <li>input: instances with only categorical features</li>
    <li>output: a list containing the most frequent feature value for each categorical feature (aka substitute values)</li>
</ul>

3. `transform` 
<ul>
    <li>input:
        <ul>
            <li>a list of instances</li>   
            <li>replacement list containing the substitute values for each feature</li>
        </ul>
    </li>
    <li>output: the transformed instances</li>
</ul>

<b> You should implement these functions from scratch yourself. You may use native Python libraries such as math, collections, or numpy to help you, but you may not use existing implementations.</b>

In [3]:
import copy

# find mean of continuous feature instances (excluding missing values)
def num_mean_fit(X):
    replacement = []
    for featureNum in range(len(X[0])):
        sumFeature = 0
        sumCounts = 0
        for featureInst in range(len(X)):
            if (X[featureInst][featureNum] != '?'):
                sumFeature += float(X[featureInst][featureNum])
                sumCounts += 1
        averageValue = sumFeature/sumCounts 
        replacement.insert(featureNum, averageValue)
    return replacement

# find mode of categorical feature instances (excluding missing values)
def cat_most_fit(X):
    replacement = []
    for featureNum in range(len(X[0])):
        categoricalFeature = []
        sumCounts = 0
        for featureInst in range(len(X)):
            if (X[featureInst][featureNum] != '?'):
                categoricalFeature.insert(sumCounts, X[featureInst][featureNum])
                sumCounts += 1
        mostCommon = max(categoricalFeature, key = categoricalFeature.count)
        replacement.insert(featureNum, mostCommon)
        categoricalFeature.clear()
    return replacement

# replace the missing values with the replacement 
def transform(X, replacement):
    trans_X  = copy.deepcopy(X)       
    
    for featureNum in range(len(trans_X[0])):
        for featureInst in range(len(trans_X)):
            if (trans_X[featureInst][featureNum] == '?'):
                trans_X[featureInst][featureNum] = replacement[featureNum]
    return trans_X


<b>For your testing:</b>

In [4]:
assert num_mean_fit([[12,122,'?'],[21,150,78],[18,'?',79]]) == [17,136,78.5]
assert cat_most_fit([['no','always'],['yes','always'],['yes','?']]) == ['yes','always']
assert transform([[10,144,'?'],[92,'?',70]], [17,136,78.5]) == [[10, 144, 78.5], [92, 136, 70]]

### 1-C: Attribute Types [1 marks]

You will create four feature representations, based on the different attribute types (categorical, continuous) in the original *Obesity* data.

**Your tasks**

Denote with $I$ the number of training instances; $C$ the number of categorical features in the data set; $N$ the number of continuous features; $v_{f}$ the number of values for categorical feature $f$; and $v_{bin}$ the number of bins for continuous features. 

1. Create a train data set with only numeric features `x_train_num` (size $(I\times N)$); and equivalently a test data set `x_test_num`; the missing values in train and test must be replaced with the average feature value found in the train dataset

2. Create a train data set with only categorical features `x_train_cat` (size $(I\times C)$); and equivalently a test data set `x_test_cat`; the missing values in train and test must be replaced with the most frequent feature values in the train dataset; You need to convert the feature values in `x_train_cat` and `x_test_cat` to integers (as shown in the Overview table) to be able to further use them in the rest of your program

3. Create a train data set with both continuous `x_train_num` and a 1-hot representation of categorical features `x_train_1hot` and save into a variable called `x_train_num_full` (size $I \times (\sum_f v_{f}+N)$) where the first $\sum_f v_f$ columns represent the 1-hot encoded categorical features and the remaining columns the continuous features; and equivalently a test data set `x_test_num_full`

4. Create a train data set using both categorical features `x_train_cat` and discretized numeric features `x_train_disc`(use equal-width binning with 4 bins, $v_{bin}=4$) and save into a variable called `x_train_cat_full` (size $I\times (C+N)$); and equivalently a test data set `x_test_cat_full`; You need to convert the feature values in `x_train_cat_full` and `x_test_cat_full` to integers (as shown in the Overview table) to be able to further use them in the rest of your program.


You will create two lists of feature names for the full categorical and full continuous datasets.

1. Create a list of feature names (string) for the `x_{train,test}_num_full` dataset (size $C+N$).

2. Create a list of feature names (string) for the `x_{train,test}_cat_full` dataset (size $\sum_f v_{f}+N$).


**Note:** You may use classes and functions from ```scikit-learn```. You may use native Python libraries such as math, collections, or numpy to help you.

In [7]:
import copy
import numpy as np
from sklearn.preprocessing import OneHotEncoder

x_train_num = []
x_test_num = []

x_train_cat = []
x_test_cat = []

x_train_num_full = []
x_test_num_full = []

x_train_cat_full = []
x_test_cat_full = []

num_feature_names = []
cat_feature_names = []

num_feature_index = []
cat_feature_index = []


# split the feature names into seperate categorical and numeric lists
def cat_num_name_split(F, f_cat, f_num, I, i_cat, i_num):
        
    numericCount = 0
    categoricalCount = 0
    numName = []
    catName = []
    # index for splitting between categorical and numeric features are defined in the overview
    for index in range(len(F)):
        if (index == 1 or index == 2 or index == 3 or index == 6 or index == 7 or index == 10 or index == 12 or index == 13):
            f_num.insert(numericCount, F[index])
            i_num.insert(numericCount, I[index])
            numericCount+=1
             
        elif (index == 0 or index == index == 4 or index == 5 or index == 8 or index == 9 or index == 11 or index == 14 or index ==15):
            f_cat.insert(categoricalCount, F[index])
            i_cat.insert(categoricalCount, I[index])
            categoricalCount+=1
              
    return
        

# split the dataset into seperate categorical and numeric lists
def cat_num_data_split(X, x_cat, x_num):
    
    # iterate through all instance rows and feature columns
    for featureInst in range(len(X)):
        numericCount = 0
        categoricalCount = 0
        numFeature = []
        catFeature = []
        
        for index in range(len(X[0])):
            if (index in num_feature_index):
                numFeature.insert(numericCount, X[featureInst][index])
                numericCount+=1
                
            elif (index in cat_feature_index):
                catFeature.insert(categoricalCount, X[featureInst][index])
                categoricalCount+=1
        
        # append the seperated values of the into a temporary list row by row into the main list, then clearing the temporary list to read next row
        x_cat.append(catFeature.copy())
        x_num.append(numFeature.copy())   
        catFeature.clear()
        numFeature.clear()
    return

        
# mapping the categorical features into defined integer feature values according to the specifications
def cat_to_integer(X):
    for instance in range(len(X)):
        for feature in range(len(X[0])):
            if (feature == 0):
                if (X[instance][feature] == "Male"):
                    X[instance][feature] = 1
                elif (X[instance][feature] == "Female"):
                    X[instance][feature] = 0
                    
            elif (feature == 1 or feature == 2 or feature == 4 or feature == 5):
                if (X[instance][feature] == "yes"):
                    X[instance][feature] = 1
                elif (X[instance][feature] == "no"):
                    X[instance][feature] = 0
                          
            elif (feature == 3 or feature == 6):
                if (X[instance][feature] == "Always"):
                    X[instance][feature] = 0
                elif (X[instance][feature] == "Frequently"):
                    X[instance][feature] = 1
                elif (X[instance][feature] == "Sometimes"):
                    X[instance][feature] = 2
                elif (X[instance][feature] == "no"):
                    X[instance][feature] = 3
                    
            else:
                if (X[instance][feature] == "Automobile"):
                    X[instance][feature] = 0
                elif (X[instance][feature] == "Bike"):
                    X[instance][feature] = 1
                elif (X[instance][feature] == "Motorbike"):
                    X[instance][feature] = 2
                elif (X[instance][feature] == "Public_Transportation"):
                    X[instance][feature] = 3
                elif (X[instance][feature] == "Walking"):
                    X[instance][feature] = 4
    return

# mapping the categorical features into one hot encoding representation                  
def cat_to_1hot_sk(x_train, x_test, cat_feature_names):
    
    # using sklearn one hot encoder to fit entire data set (since the mapping considers all possible labels)
    enc = OneHotEncoder()    
    enc.fit(x_train + x_test)
    
    # transform the training and testing set with the one hot encoder
    x_train_1hot = enc.transform(x_train)
    x_test_1hot = enc.transform(x_test)
    
    # create a list with the corresponding feature names and label
    feature_labels_1hot = []
    for feature in range(len(cat_feature_names)):
        for label in range(len(enc.categories_[feature])):
            feature_labels_1hot.append(cat_feature_names[feature] + " (" + enc.categories_[feature][label] + ")")
            
    x_train_1hot = x_train_1hot.toarray().tolist()
    x_test_1hot = x_test_1hot.toarray().tolist()
    
    return x_train_1hot, x_test_1hot, feature_labels_1hot
    
    

# converting the continuous feature instances into discretised, equal width bins
def num_to_bins(X, X_set, bins):
    
    # using numpy to transpose the list of lists to determine the minimum and maximum value
    # each feature to determine the bin width
    allFeatureColumns = np.transpose(X).astype(np.float)
    for feature in range(len(allFeatureColumns)):
        
        # finding the bin width using the entire dataset (not just testing) as this numerical to categorical
        # should be applied before the test-train split
        minFeature = (allFeatureColumns[feature]).min()
        maxFeature = (allFeatureColumns[feature]).max()
        binWidth = (maxFeature - minFeature)/bins
        
        # calculate which bin each continuous instance belongs in, note that when the instance value is equal
        # to the maximum value, it is included in the previous bin (Left bound inclusive binning).
        for instance in range(len(X_set)):
            if ((int((float(X_set[instance][feature]) - minFeature)/binWidth)) >= bins):
                X_set[instance][feature] = bins - 1
            elif ((int((float(X_set[instance][feature]) - minFeature)/binWidth)) < 0 ):
                X_set[instance][feature] = 0
            else:
                X_set[instance][feature] = (int((float(X_set[instance][feature]) - minFeature)/binWidth))   
    return
        
    
# splitting the feature names and indices
cat_num_name_split(feature_names, cat_feature_names, num_feature_names, feature_index, cat_feature_index, num_feature_index)

# splitting the dataset into categorical and numeric
cat_num_data_split(x_train, x_train_cat, x_train_num)
cat_num_data_split(x_test, x_test_cat, x_test_num)

# find the replacements for missing values in the dataset
replaceNum = num_mean_fit(x_train_num)
replaceCat = cat_most_fit(x_train_cat)

# transform the test and training data, using replacements found with the training set to replace both
x_train_num = transform(x_train_num, replaceNum)
x_train_cat = transform(x_train_cat, replaceCat)
x_test_num = transform(x_test_num, replaceNum)
x_test_cat = transform(x_test_cat, replaceCat)
x_all_num = x_train_num + x_test_num

# convert the categorical data into 1 hot encoding representation
x_train_1hot, x_test_1hot, feature_names_1hot = cat_to_1hot_sk(x_train_cat, x_test_cat, cat_feature_names)

# discretise the numeric data into 4 discretised equal-width bins
bins = 4
x_train_binning = copy.deepcopy(x_train_num)
x_test_binning = copy.deepcopy(x_test_num)
num_to_bins(x_all_num, x_train_binning, bins)
num_to_bins(x_all_num, x_test_binning, bins)

# transform categorical data into the defined integer representations
cat_to_integer(x_train_cat)
cat_to_integer(x_test_cat)

# combine the categorical and numeric lists generated into single list of lists, converting strings to floats
x_train_num = [list(map(float, sublist)) for sublist in x_train_num]
x_test_num = [list(map(float, sublist)) for sublist in x_test_num]
x_train_num_full = [x + y for x, y in zip(x_train_1hot, x_train_num)]
x_test_num_full = [x + y for x, y in zip(x_test_1hot, x_test_num)]

x_train_cat_full = [x + y for x, y in zip(x_train_cat, x_train_binning)]
x_test_cat_full = [x + y for x, y in zip(x_test_cat, x_test_binning)]

cat_feature_names_full = cat_feature_names + num_feature_names
cat_feature_index_full = cat_feature_index + num_feature_index

<b>For your testing:</b>

In [8]:
assert len(x_train_cat[0])==8
assert len(x_train_num[0])==8
assert len(x_train_cat_full[0])==16
assert len(x_train_num_full[0])==31

## Question 2: A Weighted Random baseline [1.25 marks]

Implement a <b>weighted random</b> baseline that randomly assigns a class to each test instance, weighting the class assignment according to <b>class prior probabilities</b>, as introduced in the Evaluation lecture and discuss your results.

### 2-A: Implement a weighted random baseline [0.5 marks]

**Your tasks**

You will implement the baseline function
1. `weight_random`
<ul>
  <li>input:
      <ul>
          <li> test labels</li>
          <li> the prior probability of class "obese" </li>          
      </ul>
  </li>
  <li>output: error rate on the test (fraction of incorrect predictions)</li>          
</ul> 
    

<b> You should implement this function from scratch yourself. We have already imported the two functions that we assume you will need. Do not import any other function or library.</b>

In [9]:
from collections import Counter
from random import random

test_obese = y_test.count("obese")/len(y_test) ## calculate the fraction of obese patients in the test set 
prior_obese = y_train.count("obese")/len(y_train) ## calculate the prior probability of class obese in the train set

############################
## your code begins here
############################
def weight_random(y_test, prior_obese):
    error = 0
    y_predictions = []
    
    # using prior obese probablility from training set to classify test set with random(), which generates a value between [0.0, 1.0)
    for label in range(len(y_test)):
        randomValue = random()
        if (randomValue <= prior_obese):
            y_predictions.insert(label, "obese")
        else:
            y_predictions.insert(label, "notObese")
    
    correctPredictions = 0
    for label in range(len(y_predictions)):
        if (y_test[label] == y_predictions[label]):
            correctPredictions += 1
    
    error = 1-(correctPredictions/len(y_predictions))
    
    return error 


############################
## your code ends here
############################

err = weight_random(y_test, prior_obese)
print("There are",test_obese*len(y_test),"obese and",(1-test_obese)*len(y_test),"non-obese test instances")
print("The prior probability of obese is", round(prior_obese,2))
print("The weighted random baseline's error is:", round(err,2))

##########################################
## insert code for for Question 2-B here:
##########################################

iterations = 1000
cumulativeError = 0
for i in range(iterations):
    cumulativeError += weight_random(y_test, prior_obese)
    
averageError = cumulativeError/iterations
print("\nAverage error is", averageError)

deterministicError = 1- ((prior_obese)**2+(1-prior_obese)**2)
print("Deterministic error is",deterministicError)


There are 199.0 obese and 224.0 non-obese test instances
The prior probability of obese is 0.46
The weighted random baseline's error is: 0.51

Average error is 0.49627659574468047
Deterministic error is 0.49646164169717666


### 2-B: Weighted Random baseline Discussion [0.75 marks]

Repeat the random baseline 1000 times, what would be the <b>average error rate</b> of your weighted random baseline on the test dataset? You should update your code above in order to achieve this. Please indicate through comments which parts of the code are relevant to question 2B.

What would the error rate converge to (write a formula based on the prior probability of obese, `prior_obese`, and the fraction of obese samples in the test data, `test_obese`)?

Please limit your answer to 2-3 sentences

Q2-B) Performing 1000 iterations of the weighted random baseline evaluation, the average error rate observed is approximately 0.4984. Using the deterministic estimate of accuracy for random baseline, the error rate will converge to a deterministic error of 1 - ((prior_obese)^2+(1-prior_obese)^2) = 0.4964

## Question 3: One-R baseline [2.5 marks]

In this question you will implement the one-R baseline using a helper function and discuss your results. Please refer to the Evaluation lecture for One-R baseline. 

### Question 3-A: Implement a helper function [0.75 mark]

**Your tasks**

You will implement a function that calculates the error rate for each attribute if we select that attribute as the one rule for classification.
1. `feature_error`
<ul>
  <li>input:
      <ul>
          <li> features</li>
          <li> labels</li>
      </ul>
  </li>
  <li>output:
      <ul>
          <li> a dictionary of the form {feature: error rate on the train} where each 
        feature is denoted by its ID (position in the train dataset)</li>
          <li> a dictionary of the form {feature value: majority class label} where the majority class label is either "obese" or "notObese" and feature values are the best feature's possible values (feature with smallest error rate)</li>
      </ul>
    </li>   
</ul> 

<b> You should implement this function from scratch yourself. You may use native Python libraries like math, numpy, collections to help you, but you may not use existing implementations.</b>


In [10]:
from collections import Counter
############################
## your code begins here
############################

# one r feature error for binary classification
def feature_error(features, labels):
    feat_error = {}
    feat_rule = {}
    lowestError = 1
    featureColumns =  np.transpose(features)
    classLabels = list(Counter(labels).keys())

    # count the number of times each feature label occurs
    for feature in range(len(featureColumns)):
        featureCounts = list(Counter(featureColumns[feature]).keys())
        featureValues = list(Counter(featureColumns[feature]).values())

        # threshold to find majority label greater than half
        thresholdValues = [x / 2 for x in featureValues]
        featureDict = {}
        featureDict.clear()
        errors = 0
        
        # finding the number of label matches with first classification
        for label in range(len(featureCounts)):
            matchCount = 0
            for instance in range(len(featureColumns[0])):
                if (featureColumns[feature][instance] == featureCounts[label] and labels[instance] == classLabels[0]):
                    matchCount += 1
            
            # if matches is greater than the threshold, set the feature to predict the first classification with remaining as error
            if (matchCount > thresholdValues[label]):
                dictEntry = {featureCounts[label]:classLabels[0]}
                errors += (featureValues[label] - matchCount)
                
            # if matches is less than the threshold, set the feature to predict the second classification with remaining as error
            else:
                dictEntry = {featureCounts[label]:classLabels[1]}
                errors += matchCount
            
            featureDict.update(dictEntry)
        
        # add feature error rate to dictionary
        errorRate = errors/len(featureColumns[0])
        feat_error.update({feature:errorRate})
        
        # change rule if the feature one r prediction has a lower error and update lowest error rate
        if (errorRate < lowestError):
            feat_rule = featureDict
            lowestError = errorRate
               
    return feat_error, feat_rule

############################
## your code ends here
############################

<b>For your testing</b>

In [11]:
sample_features = [[2,2], [2,1], [4,2], [4,1]]
sample_labels = ["yes","yes","no","no"]
sample_error, sample_rule= feature_error(sample_features, sample_labels)

# where index 0 refers to "feature 1" and index 1 to "feature 2"
assert sample_error[0]==0.0
assert sample_error[1]==0.5

# where 2 and 4 refer to the values of the best feature 
# and "yes","no" to the majority class label
assert sample_rule == {2:"yes",4:"no"}

### Question 3-B: Implement One-R baseline [0.75 mark]

- Apply your `feature_error` function to the full training data *x_train_cat_full*, and determine the (i) feature with *highest* error rate and (ii) feature with *lowest* error rate. Store the name (string) of the corresponding highest/lowest error rate features in `highest_err_feature_name` and `lowest_err_feature_name`, respectively.

- The feature with *lowest* error will constitute your 1-R predictor, which you should use to predict the class labels for the *x_test_cat_full* (`one_r_predictions`)


In [12]:
one_r_predictions = []

highest_err_feature_name = "" # feature name with highest error
lowest_err_feature_name = "" # feature name with lowest error

############################
## your code begins here
############################

# apply one r feature error to training set and replace the index with feature name
feature_errors, feature_rule = feature_error(x_train_cat_full, y_train)
for i in range(len(x_train_cat_full[0])):
    feature_errors[cat_feature_names_full[i]] = feature_errors.pop(i)

# acquire maximum and minimum error features
highest_err_feature_name = max(feature_errors, key=feature_errors.get)
lowest_err_feature_name = min(feature_errors, key=feature_errors.get)

# use the minimum one-r error feature weight to predict test-set
testFeatures = np.transpose(x_test_cat_full)
weightIndex = cat_feature_names_full.index("weight")
testWeightFeatures = testFeatures[weightIndex,:]  
one_r_predictions = list(map(feature_rule.get, testWeightFeatures))

############################
## your code ends here
############################

print(one_r_predictions[:10])
print(f"The feature with highest error is: {highest_err_feature_name}")
print(f"The feature with lowest error is: {lowest_err_feature_name}")


##########################################
## insert code for for Question 3-D here:
##########################################

# sort features from lowest to highest error 
sorted_one_r = dict(sorted(feature_errors.items(), key=lambda item: item[1]))
print("\nSorted One-R baseline errors:")
for key, value in sorted_one_r.items():
    print(key, ' : ', value)

['notObese', 'obese', 'notObese', 'notObese', 'notObese', 'notObese', 'obese', 'notObese', 'notObese', 'notObese']
The feature with highest error is: gender
The feature with lowest error is: weight

Sorted One-R baseline errors:
weight  :  0.2334123222748815
age  :  0.35545023696682465
family_history  :  0.3684834123222749
TUE  :  0.37618483412322273
CAEC  :  0.3921800947867299
NCP  :  0.39514218009478674
FCVC  :  0.41824644549763035
CH2O  :  0.4206161137440758
FAF  :  0.42890995260663506
height  :  0.43838862559241704
CALC  :  0.4443127962085308
FACV  :  0.4490521327014218
smoke  :  0.45734597156398105
gender  :  0.45793838862559244
SCC  :  0.45793838862559244
MTRANS  :  0.45793838862559244


### Question 3-C: Compute Mutual Information (MI) [0.5 mark]

You will calculate the mutual information between your features  *x_train_cat_full* and the class label and determine the (i) feature with *highest* MI and (ii) feature with *lowest* MI. Store the name (string) of the corresponding highest/lowest MI features in `highest_mi_feature_name` and `lowest_mi_feature_name`, respectively.

**Note:** You may use classes and functions from ```scikit-learn``` but make sure that you specify your dataset is categorical (discrete) otherwise it will be assumed that your data is numeric. 

In [13]:
from sklearn.feature_selection import mutual_info_classif
import numpy as np

highest_mi_feature_name = "" # feature with highest MI
lowest_mi_feature_name = "" # feature with lowest MI

############################
## your code begins here
############################

# apply mutual information classifier and find maximum and minimum features
mutualInfoScores = mutual_info_classif(x_train_cat_full, y_train, discrete_features=True)
highest_mi_feature_score = max(mutualInfoScores)
lowest_mi_feature_name = min(mutualInfoScores)
max_index = np.argmax(mutualInfoScores)
min_index = np.argmin(mutualInfoScores)

highest_mi_feature_name = cat_feature_names_full[max_index]
lowest_mi_feature_name = cat_feature_names_full[min_index]
############################
## your code ends here
############################

print(f"The feature with highest MI is: {highest_mi_feature_name}")
print(f"The feature with lowest MI is: {lowest_mi_feature_name}")


##########################################
## insert code for for Question 3-D here:
##########################################

# add feature name labels to mutual information score and sort from highest to lowest
mutualInfoScores = mutualInfoScores.tolist()
mutualInfoDict = {}

for i in range(len(mutualInfoScores)):
    mutualInfoDict.update({cat_feature_names_full[i]:mutualInfoScores[i]})                                                                     

sorted_MI = dict(sorted(mutualInfoDict.items(), key=lambda item: item[1], reverse=True))

print("\nSorted One-R baseline errors:")
for key, value in sorted_one_r.items():
    print(key, ' : ', value)
    
print("\nSorted MI values:")
for key, value in sorted_MI.items():
    print(key, ' : ', value)

The feature with highest MI is: weight
The feature with lowest MI is: gender

Sorted One-R baseline errors:
weight  :  0.2334123222748815
age  :  0.35545023696682465
family_history  :  0.3684834123222749
TUE  :  0.37618483412322273
CAEC  :  0.3921800947867299
NCP  :  0.39514218009478674
FCVC  :  0.41824644549763035
CH2O  :  0.4206161137440758
FAF  :  0.42890995260663506
height  :  0.43838862559241704
CALC  :  0.4443127962085308
FACV  :  0.4490521327014218
smoke  :  0.45734597156398105
gender  :  0.45793838862559244
SCC  :  0.45793838862559244
MTRANS  :  0.45793838862559244

Sorted MI values:
weight  :  0.35716955764230074
family_history  :  0.10911015300052182
CAEC  :  0.08540000889807925
NCP  :  0.057309778401144915
TUE  :  0.04385790084261236
FACV  :  0.041641430427852386
age  :  0.04017873112638997
SCC  :  0.022348931879498865
height  :  0.01586715905733986
FAF  :  0.01432647846253706
CALC  :  0.0140082708635033
FCVC  :  0.01335145986162388
MTRANS  :  0.013265084696444928
CH2O  :  0

### Question 3-D: Discussion of One-R and Mutual Information [0.5 mark]
In addition to features leading to lowest/highest error rates and MI, inspect the MI and error rates for all 16 features. You should update your code above in order to achieve this. Please indicate through comments which parts of the code are relevant to question 3D.  

Are you finding different or somewhat similar patterns in the results? Explain the difference and similarity between MI and error rate formula in One-R?

Please limit your answer to 2-3 sentences

Q3-D) There is a general pattern when comparing the MI and one-R baseline feature error rate, where features with a low MI score has a high error rate, with weight and family history both having low error and corresponding high MI, however, there is a difference with age, where it has the 2nd lowest error rate but only the 7th highest MI value. Also, there is a difference with the magnitude of variation, where the one-R error rate for a particular feature converging to a maximum of the prior obese probability because it takes the majority class label for each feature value, causing the gender, SCC and MTRANS to have the same error rate. On the other hand, the MI scores have different values for these 3 features with gender having a much lower score of 5.82e-5 as compared to SCC with 2.23e-3, indicating that gender is less correlated to obesity compared to SCC according to MI.

## Question 4: Naive Bayes [2 marks]

We will construct two Naive Bayes classifiers

1. One for instances with categorical attributes.
2. One for instances with numerical attributes.


For each classifier, you will 
1. Train it on the training set
2. Use the models to predict labels for the test set

### Question 4-A: Implementing the Naive Bayes classifiers [0.5 marks]

Implement two functions which train a NB classifier given the specified input feature types and predict labels for a given test set.

You may add additional quantities to your `return` statements.

<b> You may (and are, indeed, encouraged) to, use the existing NB implementations from `sklearn`. You should use the default parameterizations of these algorithms.</b> 
    
**Note:** If you choose to implement your classifiers from scratch, please use Laplace smoothing (alpha=1) for the categorical feature NB.




In [14]:
from sklearn.naive_bayes import CategoricalNB
from sklearn.naive_bayes import GaussianNB

##################################
## your code begins here
##################################

# using Categorical NB model and fitting train instances. Apply model on test features 
def nb_cat_features(train_features, train_labels, test_features):
    predictions = []      
    classifier = CategoricalNB()
    classifier.fit(train_features, train_labels)
    predictions = classifier.predict(test_features)
    return predictions

# using Gaussian NB model and fitting train instances. Apply model on test features 
def nb_num_features(train_features, train_labels, test_features):
    predictions = []      
    classifier = GaussianNB()
    classifier.fit(train_features, train_labels)
    predictions = classifier.predict(test_features)
    return predictions

############################
## your code ends here
############################

### Question 4-B: Apply your classifiers to the data sets you created in Questions 1 [0.5 marks]

...namely, the

1. categorical features `x_{train,test}_cat`
2. numerical features `x_{train,test}_num`
3. combined discretized numerical and categorical features `x_{train,test}_cat_full`
4. combined 1-hot categorical features and numerical `x_{train,test}_num_full` 


In [15]:
from collections import Counter

############################
## your code begins here
############################

cat_nb_predictions = nb_cat_features(x_train_cat, y_train, x_test_cat) # predictions for x_test_cat
num_nb_predictions = nb_num_features(x_train_num, y_train, x_test_num) # predictions for x_test_num
full_cat_nb_predictions = nb_cat_features(x_train_cat_full, y_train, x_test_cat_full) # predictions for x_test_cat_full
full_num_nb_predictions = nb_num_features(x_train_num_full, y_train, x_test_num_full) # predictions for x_test_num_full

############################
## your code ends here
############################

print(f"Numerical NB predicted class distribution \t {Counter(num_nb_predictions)}")
print(f"Categorical NB predicted class distribution\t {Counter(cat_nb_predictions)}")
print(f"Full Categorical NB predicted class distribution\t {Counter(full_cat_nb_predictions)}")
print(f"Full Numerical NB predicted class distribution\t {Counter(full_num_nb_predictions)}")

Numerical NB predicted class distribution 	 Counter({'notObese': 237, 'obese': 186})
Categorical NB predicted class distribution	 Counter({'obese': 261, 'notObese': 162})
Full Categorical NB predicted class distribution	 Counter({'notObese': 215, 'obese': 208})
Full Numerical NB predicted class distribution	 Counter({'obese': 261, 'notObese': 162})


### Question 4-C: Combining NB predictions [1 mark]

You are asked to construct another Naive Bayes classifier for instances with the full set of numerical *and* categorical attributes (no discretization and no one-hot encoding), ensuring that your classifier computes posterior class probabilities $p(y|x)$ as $p(x|y)p(y)$. Explain such Naive Bayes classifier implementation on the full attribute set (numerical and categorical features) 

Please limit your answer to 3-4 sentences

Q4-C) The classifier will begin with calculating the prior class probabilities for obese and notObese classifications p(y) using a bernoulli distribution as it has binary class labels. After this, the posterior class probabilities for each feature 𝑝(𝑦|𝑥) will be calculated, where 𝑝(𝑥|𝑦) is determined using a gaussian distribution for numerical features with mean μ and standard deviation σ. For categorical features, 𝑝(𝑥|𝑦) is detemined using a categorical distribution, calculated by taking the observed relative frequency of (class, label) divided by the total frequency of that label. Finally, the the 𝑝(𝑦|𝑥) is determined with $p(y)\prod_{i=0}^{n} p(x_n|y)$ for n features and the class label the NB model predicts is found using $\underset{y\in Y}{\mathrm{argmax}} p(y)\prod_{i=0}^{n} p(x_n|y)$, or equivalently, the log transformation $\underset{y\in Y}{\mathrm{argmax}} [\log(p(y)) + \sum_{i=0}^{n} \log(p(x_n|y))]$ to avoid underflow with floating-point number representation.

## Question 5: Logistic Regression [1.25 marks]

1. First find the mean and standard deviation of values for each feature in `x_train_num_full` and then using the calculated mean and standard deviation values, standardize `x_{train,test}_num_full` such that each feature has mean=0 and std= 1. Store the standardized data to `x_{train,test}_norm_full`.

2. Implement a Logistic Regression classifier, and apply it to the standardized full numeric training data set (`x_{train,test}_norm_full`).

<b> You should use existing implementation in sklearn with default parameters for Logistic Regression. You may use classes and functions from ```scikit-learn``` for standardization.</b>

### Question 5-A: The Logistic Regression classifier [0.5 marks]


In [14]:
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler

lr_predictions = []
x_train_norm_full = []
x_test_norm_full = []
lr_coeffs = {}

############################
## your code begins here
############################

# fit standard scaler with test and apply to numerical features (not one-hot encoded) using standard scaler (mean 0, stddev 1)
scaler = StandardScaler()
scaler.fit(x_train_num)
x_train_norm = scaler.transform(x_train_num).tolist()
x_test_norm = scaler.transform(x_test_num).tolist()
x_train_norm_full = np.array([x + y for x, y in zip(x_train_1hot, x_train_norm)])
x_test_norm_full = np.array([x + y for x, y in zip(x_test_1hot, x_test_norm)])

# fit logistic regression model with training instances then apply on testing instances
classifier = LogisticRegression().fit(x_train_norm_full, y_train)
lr_predictions = classifier.predict(x_test_norm_full)

############################
## your code ends here
############################

print(f"LR predicted class distribution\t {Counter(lr_predictions)}")

##########################################
## insert code for for Question 5-B here:
##########################################
num_full_feature_names = feature_names_1hot + num_feature_names
compare_coeffs_MI = {}

# adding labels to coefficients and sorting them from largest to smallest magnitude
for i in range(len(num_full_feature_names)):
    lr_coeffs.update({num_full_feature_names[i]:classifier.coef_[0][i]}) 
lr_coeffs

sorted_coeffs = dict(sorted(lr_coeffs.items(), key=lambda item: abs(item[1]), reverse=True))

print("\nSorted MI values:")
for key, value in sorted_MI.items():
    print(key, ' : ', value)
print("\nSorted LR coefficients by magnitude:")
for key, value in sorted_coeffs.items():
    print(key, ' : ', value)

LR predicted class distribution	 Counter({'notObese': 226, 'obese': 197})

Sorted MI values:
weight  :  0.35716955764230074
family_history  :  0.10911015300052182
CAEC  :  0.08540000889807925
NCP  :  0.057309778401144915
TUE  :  0.04385790084261236
FACV  :  0.041641430427852386
age  :  0.04017873112638997
SCC  :  0.022348931879498865
height  :  0.01586715905733986
FAF  :  0.01432647846253706
CALC  :  0.0140082708635033
FCVC  :  0.01335145986162388
MTRANS  :  0.013265084696444928
CH2O  :  0.012763756202343879
smoke  :  0.00014290753095302228
gender  :  5.822429487567504e-05

Sorted LR coefficients by magnitude:
weight  :  6.29373183935817
height  :  -1.622265895807806
CAEC (Sometimes)  :  1.2131569215413787
CAEC (Frequently)  :  -1.0865002090013596
family_history (yes)  :  0.9616974108641757
family_history (no)  :  -0.9607784099945917
SCC (no)  :  0.8359092825544131
SCC (yes)  :  -0.8349902816848271
FACV (yes)  :  0.7618067986006659
FACV (no)  :  -0.7608877977310814
MTRANS (Walking)  : 

### Question 5-B: Inspect the learned coefficients [0.75 marks]

**Part 1**
Inspect the coefficients (weights) learned by the classifier, and compare them to the features with the highest MI values in Question 3. You should update your code above in order to achieve this. Please indicate through comments which parts of the code are relevant to question 5B.

**Part 2**
What does a high coefficient for a feature imply? Report the two feature names for which your model learns the most important coefficients? Compare your answer to the highest MI features you detected in Q3.


Q5-B)The features with the highest 3 MI values are weight, family history and CAEC; the features with the highest 3 LR coefficients are weight, height and CAEC (Frequently). The weight feature has the highest MI value and corresponding highest magnitude LR coefficient, however, there is a clear difference in the general ordering, with an example being the height coefficient having the second largest LR coefficient magnitude but only the 9th highest MI score, which could be due to MI losing information from discretised binning, and gender having the lowest MI score but a LR coefficient roughly halfway in the ordered magnitude of coefficients.

A high coefficient (magnitude-wise) implies that the feature is highly correlated with the classification outcome, where a positive coefficient representing a positive correlation and a negative coefficient representing a negative correlation. The 2 most important coefficient are weight and height, where weight is positively correlated with an obese class label (6.4) and height being negatively correlated to the obese class label (-1.68). Comparing this with MI scores, weight has a corresponding highest MI score (0.36), however, height only has the 9th highest MI score (0.016).

## Question 6: Evaluation [1 mark]

We will evaluate our baselines and classifiers on the instances in the test set. 

Compute 
- error rate
- macro-averaged F1 score 

for the One-R baselines, the four NB classifiers (on 4 datasets) and the LR classifier.

**You may use existing implementations and/or Python libraries like numpy, scipy or sklearn.**

In [15]:
from sklearn.metrics import f1_score
from sklearn.metrics import accuracy_score

one_r_err = 1 - accuracy_score(one_r_predictions, y_test)
one_r_f1 = f1_score(one_r_predictions, y_test, average='macro')


num_nb_err = 1 - accuracy_score(num_nb_predictions, y_test)
num_nb_f1 = f1_score(num_nb_predictions, y_test, average='macro')


cat_nb_err = 1 - accuracy_score(cat_nb_predictions, y_test)
cat_nb_f1 = f1_score(cat_nb_predictions, y_test, average='macro')


full_cat_nb_err = 1 - accuracy_score(full_cat_nb_predictions, y_test)
full_cat_nb_f1 = f1_score(full_cat_nb_predictions, y_test, average='macro')


full_num_nb_err = 1 - accuracy_score(full_num_nb_predictions, y_test)
full_num_nb_f1 = f1_score(full_num_nb_predictions, y_test, average='macro')

lr_err = 1 - accuracy_score(lr_predictions, y_test)
lr_f1 = f1_score(lr_predictions, y_test, average='macro')


print(f"One  R\t\tError: {round(one_r_err, 2)}\tMacro F1: {round(one_r_f1, 2)}")
print(f"NB Num \t\tError: {round(num_nb_err, 2)}\tMacro F1: {round(num_nb_f1, 2)}")
print(f"NB Cat \t\tError: {round(cat_nb_err, 2)}\tMacro F1: {round(cat_nb_f1, 2)}")
print(f"NB Num Full \tError: {round(full_num_nb_err, 2)}\tMacro F1: {round(full_num_nb_f1, 2)}")
print(f"NB Cat Full \tError: {round(full_cat_nb_err, 2)}\tMacro F1: {round(full_cat_nb_f1, 2)}")
print(f"LR \t\tError: {round(lr_err, 2)}\tMacro F1: {round(lr_f1, 2)}")

One  R		Error: 0.23	Macro F1: 0.74
NB Num 		Error: 0.13	Macro F1: 0.87
NB Cat 		Error: 0.21	Macro F1: 0.79
NB Num Full 	Error: 0.21	Macro F1: 0.79
NB Cat Full 	Error: 0.19	Macro F1: 0.81
LR 		Error: 0.06	Macro F1: 0.94


## Question 7: Discussion [5 marks] 

Critically analyze the performance of the models by answering the following questions. 

**(a)** The four Naive Bayes (NB) classifiers lead to different performance. Which of the four NB classifiers performs best, and why do you think it is the case? **[1 mark]**

**(b)** You used unstandardized data in your numeric Naive Bayes, what would be the effect of standardization on the results of numeric Naive Bayes? Why? **[0.75 mark]**

**(c)** You chose to use equal-width binning for your continuous features, discuss the drawbacks of considering this approach with a fixed bin number. Would you select a different binning strategy, and if so which? **[1 mark]**

**(d)** Assume that your best classifier will be deployed in a hospital to warn new patients if they are prone to obesity. You want to avoid falsely diagnosing a patient as obese. Can you reassure the hospital with the evaluation results in Q7 that your classifier is adequate? If not, describe an alternative evaluation metric and explain why it is more appropriate.**[0.75 mark]**

**(e)** Do you observe a clear difference in performance between NB and Logistic Regression (LR) in Q7? Referring to the number of parameters and assumptions underlying each model, provide one reason why NB might outperform LR, and one reason why LR might outperform NB. In your answer, refer back to the *Obesity* data set used in this assignment. **[1 mark]**

**(f)** Assume that you have access to a (hypothetical) infinitely large database (i.e., an enormous version of the Obesity data set used in this assignment). Is it true that Logistic Regression will achieve perfect test set performance? Why (not)? **[0.5 mark]**


<b>We expect a maximum of 2-3 sentences per question.

(a) The Gaussian Naive Bayes on the just the numerical features leads to the best performance with an error rate of 0.13 and macro F1 score of 0.87. Comparing with the Gaussian Naive Bayes on the one hot encoded categorical features + numerical features, the error rate from this method is higher at 0.21, which could be due to applying a gaussian distribution to binary feature labels, where a bernoulli distribution should be chosen instead. Furthermore, comparing with the categorical naive bayes sets, the error rate from this method is higher at 0.19 and 0.21 for the discretised full and only categorical features respectively, which could be due to information loss of numerical features from discretised binning and using categorical distribution on binary features, where a bernoulli distribution should be used instead.


(b) Standardisation on the numerical data before performing Gaussian Naive Bayes will not lead to a change in the results because standardising the numerical features to a mean of 0 and standard deviation of 1 will scale the numerical values to a normal distribution. This procedure of fitting the numerical features into a normal distribution is a step performed in the Gaussian Naive Bayes, therefore, applying this standardisation before performing Gaussian Naive Bayes will not change the results.


(c) The drawback with fixing the number of equal-width bins for all numerical features is that the range of the numerical features vary greatly. This means that numerical features with a large range such as weight will lose information due to a large bin width, and features with a small range such as TUE could create excessive bins with a small bin width which does not give a sense of distribution. The binning strategy I would use is an equal bin width but using the Freedman–Diaconis rule to create a different number of bins depending on the numerical feature, to better characterise the distribution $\text{Bin Width} = 2\frac{IQR}{\sqrt[3]n}$.


(d) Since it is important to avoid falsely diagnosing a patient as obese, false positives have to be low compared to the number of false negatives. Therefore, precision would be a the important measure to reduce the number of false positives instead of error rate, which could be misleadingly high with a high true negative rate but precision will not reveal problems with an algorithm that predicts every patient as notObese. Accounting for this, a F0.5 scores is used to balance recall and precision, where precision is emphasised over recall to ensure a low false postive rate while to a lesser extent, accounting for false negatives to avoid diagnosing obese patients as not obese.


(e) There is a clear difference in error rate when comparing the discriminative logistic regression and the generative naive bayes classifiers, with logistic regression having a much lower error of 0.06 compared to the best naive bayes classifier error of 0.13. LR outperforms NB because the dataset is large enough (1688 training instances) and LR maximises the posterior probability directly by minimising the negative conditional log likelihood, however, NB assumes conditional independance between the features to arrive at the posterior probability even though there are some features that are highly dependent such as height and weight. NB will outperform LR if the dataset is small and there are limited training instances because of the independence assumption which allows the model to converge quicker, as compared to LR which needs a large amount of data converge.


(f) An infinitely large database will not lead to perfect test set performance because (assuming the features are linearly seperable), using iterative gradient descent on the derivative of the log likelihood for each parameter will converge to a global minimum with log-likelihood being globally concave. However, this does not mean that the LR coefficients are perfect for every test set as it is fit to a training set, and far outliers in the test set or test instances where feature values are identical to instances in the training set with a different classification can occur in real-life datasets.
